citar-cache.el

```diff
@@ -0,0 +1,239 @@
+is nil, use the current buffer.  Otherwise, BUFFER should be a
+buffer object or name that requires the bibliography FILENAME, or
+a symbol like 'global."
+  (let* ((cached (gethash filename citar-cache--bibliographies))
+         (bib (or cached (citar-cache--make-bibliography filename)))
+         (buffer (citar-cache--canonicalize-buffer buffer))
+         (fmtstr (citar--get-template 'completion)))
+    (unless cached
+      (setf (citar-cache--bibliography-format-string bib) fmtstr)
+      (citar-cache--update-bibliography bib)
+      (puthash filename bib citar-cache--bibliographies))
+    ;; Preformat strings if format has changed
+    (unless (equal-including-properties
+             fmtstr (citar-cache--bibliography-format-string bib))
+      (setf (citar-cache--bibliography-format-string bib) fmtstr)
+      (citar-cache--preformat-bibliography bib))
+    ;; Associate buffer with this bibliography:
+    (cl-pushnew buffer (citar-cache--bibliography-buffers bib))
+    ;; Release bibliography when buffer is killed or changes major mode:
+    (unless (symbolp buffer)
+      (with-current-buffer buffer
+        (dolist (hook '(change-major-mode-hook kill-buffer-hook))
+          (add-hook hook #'citar-cache--release-bibliographies 0 'local))))
+    bib))
+
+
+(defun citar-cache--release-bibliographies (&optional keep-filenames buffer)
+  "Dissociate BUFFER from cached bibliographies.
+If BUFFER is nil, use the current buffer.  Otherwise, BUFFER
+should be a buffer object, buffer name, or a symbol like 'global.
+KEEP-FILENAMES is a list of file names that are not dissociated
+from BUFFER.
+
+Remove any bibliographies from the cache that are no longer
+needed by any other buffer."
+  (let ((buffer (citar-cache--canonicalize-buffer buffer)))
+    (maphash
+     (lambda (filename bib)
+       (unless (member filename keep-filenames)
+         (cl-callf2 delq buffer (citar-cache--bibliography-buffers bib))
+         (unless (citar-cache--bibliography-buffers bib)
+           (citar-cache--remove-bibliography filename))))
+     citar-cache--bibliographies)))
+
+
+(defun citar-cache--remove-bibliography (filename)
+  "Remove bibliography cache entry for FILENAME."
+  ;; TODO Perform other needed  actions, like removing filenotify watches
+  (remhash filename citar-cache--bibliographies))
+
+
+;;; Updating bibliographies
+
+
+(defun citar-cache--update-bibliography (bib &optional force)
+  "Update the bibliography BIB from the original file.
+
+Unless FORCE is non-nil, the file is re-read only if it has been
+modified since the last time BIB was updated."
+  (let* ((filename (citar-cache--bibliography-filename bib))
+         (entries (citar-cache--bibliography-entries bib))
+         (newhash (with-temp-buffer
+                    (insert-file-contents filename)
+                    (buffer-hash))))
+    ;; TODO Also check file size and modification time before hashing?
+    ;; See `file-has-changed-p` in emacs 29, or `org-file-has-changed-p`
```

Comment:
I didn't realize this was a new function, but below is the source code. I think the org version wraps this, with some backward compatibility stuff.

So it doesn't look at a checksum, but at file size and modification time.

```elisp
(defun file-has-changed-p (file &optional tag)
  "Return non-nil if FILE has changed.
The size and modification time of FILE are compared to the size
and modification time of the same FILE during a previous
invocation of `file-has-changed-p'.  Thus, the first invocation
of `file-has-changed-p' always returns non-nil when FILE exists.
The optional argument TAG, which must be a symbol, can be used to
limit the comparison to invocations with identical tags; it can be
the symbol of the calling function, for example."
  (let* ((file (directory-file-name (expand-file-name file)))
         (remote-file-name-inhibit-cache t)
         (fileattr (file-attributes file 'integer))
	 (attr (and fileattr
                    (cons (file-attribute-size fileattr)
		          (file-attribute-modification-time fileattr))))
	 (sym (concat (symbol-name tag) "@" file))
	 (cachedattr (gethash sym file-has-changed-p--hash-table)))
     (when (not (equal attr cachedattr))
       (puthash sym attr file-has-changed-p--hash-table))))
```